### PR TITLE
drivers: pinctrl: Add ZynqMP / Mercury XU pinctrl support

### LINKED ISSUE #52437: drivers: pinctrl: add pinctrl driver for the Xilinx ZynqMP platform
**Is your enhancement proposal related to a problem? Please describe.**
As noted in #39740 the Xilinx ZynqMP platform uses a custom pinctrl driver (as seen in https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/arm/mercury_xu/board.c).

**Describe the solution you'd like**
The Xilinx ZynqMP platform should use one of the standard Zephyr pinctrl devicetree representations as described in https://docs.zephyrproject.org/latest/hardware/pinctrl/index.html#devicetree-representation along with a pinctrl driver.

**Describe alternatives you've considered**
None.

**Additional context**
See the Xilinx Zynq-7000 series pinctrl binding + driver for inspiration.

### PR DESCRIPTION
**Pull request by jbylicki**
This PR removes the custom pin control setup from the `board.c` file for the Mercury XU platform and replaces it with an implementation compliant with the `pinctrl` specification.

Fixes #52437

### CODE REVIEW

File: drivers/pinctrl/pinctrl_xlnx_zynqmp.c [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/drivers/syscon.h>
+#include <zephyr/devicetree.h>
+
+LOG_MODULE_REGISTER(pinctrl_xlnx_zynqmp, CONFIG_PINCTRL_LOG_LEVEL);
+
+#define DT_DRV_COMPAT xlnx_pinctrl_zynqmp
+
+BUILD_ASSERT(DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT) == 1, "Unsupported number of instances");
```
Comment (gmarull) [inline drivers/pinctrl/pinctrl_xlnx_zynqmp.c:20]: redundant

File: drivers/pinctrl/pinctrl_xlnx_zynqmp.c [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+ */
+
+#include "pinctrl_soc.h"
+#include <zephyr/dt-bindings/pinctrl/pinctrl-zynqmp.h>
+#include <zephyr/device.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/drivers/syscon.h>
+#include <zephyr/devicetree.h>
```
Comment (gmarull) [inline drivers/pinctrl/pinctrl_xlnx_zynqmp.c:14]: please, cleanup includes. not everything looks necessary

File: drivers/pinctrl/pinctrl_xlnx_zynqmp.c [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+LOG_MODULE_REGISTER(pinctrl_xlnx_zynqmp, CONFIG_PINCTRL_LOG_LEVEL);
+
+#define DT_DRV_COMPAT xlnx_pinctrl_zynqmp
```
Comment (gmarull) [inline drivers/pinctrl/pinctrl_xlnx_zynqmp.c:18]: driver does not use any DT APIs, other than the redundant line below.

File: dts/bindings/pinctrl/xlnx,pinctrl-zynqmp.yaml [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+include: base.yaml
+
+child-binding:
+  description: |
+    Definitions for a pinctrl state.
+  child-binding:
+
+    include:
+      - name: pincfg-node.yaml
+        property-allowlist:
```
Comment (gmarull) [inline dts/bindings/pinctrl/xlnx,pinctrl-zynqmp.yaml:21]: if not allowing anything, just skip this

File: include/zephyr/dt-bindings/pinctrl/pinctrl-zynqmp.h [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_DT_BINDINGS_PINCTRL_ZYNQMP_PINCTRL_H_
+#define ZEPHYR_INCLUDE_DT_BINDINGS_PINCTRL_ZYNQMP_PINCTRL_H_
+
+#define UART_FUNCTION 0x1
+
+#define UART0_RX_38 (38U | (UART_FUNCTION << 8))
+#define UART0_TX_39 (39U | (UART_FUNCTION << 8))
```
Comment (gmarull) [inline include/zephyr/dt-bindings/pinctrl/pinctrl-zynqmp.h]: define magic 8?

File: soc/arm/xilinx_zynqmp/pinctrl_soc.h [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+#ifndef ZEPHYR_SOC_ARM_XLNX_ZYNQMP_SOC_PINCTRL_H_
+#define ZEPHYR_SOC_ARM_XLNX_ZYNQMP_SOC_PINCTRL_H_
+
+#include <zephyr/devicetree.h>
+#include <zephyr/sys/util.h>
+#include <zephyr/types.h>
+
+#define MIO_DEFAULT	   0x0
+#define MIO_BASE_ADDR  0xff180000
+#define MIO_PIN_OFFSET 0x04
```
Comment (gmarull) [inline soc/arm/xilinx_zynqmp/pinctrl_soc.h:16]: please no addresses out of DT, is this even needed?

File: soc/arm/xilinx_zynqmp/pinctrl_soc.h [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+#define MIO_L0_SEL     BIT(1)
+#define MIO_L1_SEL     BIT(2)
```
Comment (gmarull) [inline soc/arm/xilinx_zynqmp/pinctrl_soc.h:19]: not used?

**Review by gmarull — CHANGES_REQUESTED** [review]
please split board changes from driver commit

File: drivers/pinctrl/pinctrl_xlnx_zynqmp.c (hunk 3 above)
Comment (jbylicki) [reply drivers/pinctrl/pinctrl_xlnx_zynqmp.c:13]: Now it does use DT API to fetch the MIO Register base address. Should it be removed despite that?

File: soc/arm/xilinx_zynqmp/pinctrl_soc.h (hunk 7 above)
Comment (jbylicki) [reply soc/arm/xilinx_zynqmp/pinctrl_soc.h:15]: Yes, that's unused, but I think it should be left here. It defines the layout of the register. One less thing to check for the next person using it.

Comment (gmarull) [reply soc/arm/xilinx_zynqmp/pinctrl_soc.h:15]: dead code is not allowed in Zephyr

Comment (henrikbrixandersen) [reply soc/arm/xilinx_zynqmp/pinctrl_soc.h:15]: I don't think this qualifies as dead code. A lot of device drivers and SoCs define all of their registers and fields, even if they are not used yet.

Comment (fkokosinski) [reply soc/arm/xilinx_zynqmp/pinctrl_soc.h:15]: Hi @henrikbrixandersen, @gmarull! It looks like the discussion here has been stalled for some time now.

We’ve originally added the `MIO_L1_SEL` field definition for reasons stated by @henrikbrixandersen, but we’re okay with either way - keeping this define or removing it altogether.

Let us know how we should proceed with this. Thanks!

Comment (henrikbrixandersen) [reply soc/arm/xilinx_zynqmp/pinctrl_soc.h:15]: I'd keep it.

Comment (gmarull) [reply soc/arm/xilinx_zynqmp/pinctrl_soc.h:15]: Ok, it won't hurt, it's just a single definition with potential future use within pinctrl_soc.h.

**Review by gmarull — APPROVED** [review]

**Review by henrikbrixandersen — APPROVED** [review]

**Review by fkokosinski — APPROVED** [review]

**Pull request merged**